data/transactions/application.go

```diff
@@ -416,65 +416,181 @@ func (ac *ApplicationCallTxnFields) AddressByIndex(accountIdx uint64, sender bas
-		record.TotalAppSchema = totalSchema
+	// Update the cached TotalStateSchema for this account, used
+	// when computing MinBalance, since the creator has to store
+	// the global state
+	totalSchema := record.TotalAppSchema
+	totalSchema = totalSchema.AddSchema(ac.GlobalStateSchema)
+	record.TotalAppSchema = totalSchema
 
-		// Tell the cow what app we created
-		created := []basics.CreatableLocator{
-			basics.CreatableLocator{
-				Creator: header.Sender,
-				Type:    basics.AppCreatable,
-				Index:   basics.CreatableIndex(appIdx),
-			},
-		}
+	// Tell the cow what app we created
+	created := []basics.CreatableLocator{
+		{
+			Creator: creator,
+			Type:    basics.AppCreatable,
+			Index:   basics.CreatableIndex(appIdx),
+		},
+	}
 
-		// Write back to the creator's balance record and continue
-		err = balances.PutWithCreatables(record, created, nil)
-		if err != nil {
-			return err
-		}
+	// Write back to the creator's balance record and continue
+	err = balances.PutWithCreatables(record, created, nil)
+	if err != nil {
+		return 0, err
 	}
 
+	return
+}
+
+func (ac *ApplicationCallTxnFields) applyClearState(
+	balances Balances, sender basics.Address, appIdx basics.AppIndex,
+	ad *ApplyData, steva StateEvaluator,
+) error {
+
 	// Fetch the application parameters, if they exist
 	params, creator, exists, err := getAppParams(balances, appIdx)
 	if err != nil {
 		return err
 	}
 
+	record, err := balances.Get(sender, false)
+	if err != nil {
+		return err
+	}
+
+	// Ensure sender actually has LocalState allocated for this app.
+	// Can't clear out if not currently opted in
+	_, ok := record.AppLocalStates[appIdx]
+	if !ok {
+		return fmt.Errorf("cannot clear state for app %d: account %s is not currently opted in", appIdx, sender.String())
+	}
+
+	// If the application still exists...
+	if exists {
+		// Execute the ClearStateProgram before we've deleted the LocalState
+		// for this account. If the ClearStateProgram does not fail, apply any
+		// state deltas it generated.
+		pass, stateDeltas, err := steva.Eval(params.ClearStateProgram)
+		if err == nil && pass {
+			// Program execution may produce some GlobalState and LocalState
+			// deltas. Apply them, provided they don't exceed the bounds set by
+			// the GlobalStateSchema and LocalStateSchema. If they do exceed
+			// those bounds, then don't fail, but also don't apply the changes.
+			failIfNotApplied := false
+			err = ac.applyStateDeltas(stateDeltas, params, creator, sender,
+				balances, appIdx, failIfNotApplied)
+			if err != nil {
+				return err
+			}
+
+			// Fill in applyData, so that consumers don't have to implement a
+			// stateful TEAL interpreter to apply state changes
+			ad.EvalDelta = stateDeltas
+		} else {
+			// Ignore errors and rejections from the ClearStateProgram
+		}
+	}
+
+	// Fetch the (potentially updated) sender record
+	record, err = balances.Get(sender, false)
+	if err != nil {
+		return err
```

Comment:
To be consistent, since you moved the `ad.EvalDelta` in `apply`, maybe we should set `ad.EvalDelta = nil` here, and add error checking to the final `balances.Put` below, and set it `nil` in the error case as well.

Reply:
done